torch/distributed/_shard/checkpoint/filesystem.py

```diff
@@ -0,0 +1,127 @@
+import io
+import os
+import operator
+import pickle
+from typing import List, Optional, cast
+
+import torch
+from torch import Tensor
+import torch.distributed as dist
+from torch.futures import Future
+
+from .metadata import (
+    BytesReadRequest,
+    BytesWriteRequest,
+    Metadata,
+    TensorReadRequest,
+    TensorWriteRequest,
+)
+from .storage import StorageReader, StorageWriter
+
+class FileSystemWriter(StorageWriter):
+    def __init__(self, path: str) -> None:
+        super().__init__()
+        self.path = path
+        os.makedirs(self.path, exist_ok=True)
+
+    def write_bytes(self, requests: List[BytesWriteRequest]) -> Future[None]:
+        for req in requests:
+            with open(os.path.join(self.path, req.storage_key), "wb") as storage:
+                storage.write(req.bytes.getbuffer())
+
+        fut: Future[None] = Future()
+        fut.set_result(None)
+        return fut
+
+    def write_tensors(self, requests: List[TensorWriteRequest]) -> Future[None]:
+        for req in requests:
+            with open(os.path.join(self.path, req.storage_key), "wb") as storage:
+                # The following couple lines are simple implementation to get
+                # things going.
+                #
+                # At load time, to enable resharding, we use (sub)view of the tensor.
+                # Since the storage of the tensor might not be contiguous. we need to
+                # preseve the original view, to calculate the correct sub view at load.
+                #
+                # `torch.save` saves both the view and storage, it is a good option
+                # for unblocking. There are two drawbacks:
+                # 1. `torch.save` is pickle based, and pickle is not known for its
+                #   compatibility, we should consider replacing it with a more
+                #   stable option.
+                # 2. pickle is not streamable.
+                buffer = io.BytesIO()
+                torch.save(req.tensor, buffer)
+                storage.write(buffer.getbuffer())
+
+        fut: Future[None] = Future()
+        fut.set_result(None)
+        return fut
+
+    # Implementating the abstract function in Storage Writer
+    def write_metadata(self, metadata: Metadata) -> None:
+        # Once write metadata once as each Metadata has the global view
+        if dist.is_initialized() and dist.get_rank() != 0:
+            return
+
+        with open(os.path.join(self.path, ".metadata"), "wb") as metadata_file:
+            pickle.dump(metadata, metadata_file)
```

Comment:
We're using `torch.save` for saving individual entries, which uses pickle as well, that's another problematic place.

We're considering switching to flatbuffers, but the discussion still going on.

Reply:
For tensors I think we need to write out the tensor contents directly.  We want to minimize the I/O overhead here, and therefore don't want to go through a serialization layer that might force copying the data.  For tensors stored in GPU we also should write out the data in a streaming fashion: we ideally should not have to hold the entire tensor data in CPU memory at once while saving the checkpoint.

For tensor metadata (i.e., the tensor options) then we will need some serialization layer to save this data.  I would perhaps be more inclined to do this serialization in C++ rather than in Python.